arangod/Aql/OptimizerRules.cpp

```diff
@@ -2044,23 +2044,59 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
     // finally, adjust the original plan and create a sorted version of COLLECT.
     collectNode->aggregationMethod(CollectOptions::CollectMethod::kSorted);
 
-    // insert a SortNode IN FRONT OF the CollectNode
+    // insert a SortNode IN FRONT OF the CollectNode, if we don't already have
+    // one that can be used instead
     if (!groupVariables.empty()) {
-      SortElementVector sortElements;
-      for (auto const& v : groupVariables) {
-        sortElements.push_back(SortElement::create(v.inVar, true));
+      // check if our input is already sorted
+      SortNode* sortNode = nullptr;
+      auto d = n->getFirstDependency();
+      while (d) {
+        if (d->getType() == EN::SORT) {
+          sortNode = ExecutionNode::castTo<SortNode*>(d);
+          break;
+        }
+        if (d->getType() == EN::INDEX ||
```

Comment:
I'd prefer a positive list of nodes that are allowed to be between the collect and sort. This list already missing the join node.